lib/spack/spack/relocate.py

```diff
@@ -0,0 +1,210 @@
+            tty.warn(output)
+
+    for orig, new in zip(rpaths, nrpaths):
+        command= ("install_name_tool -rpath %s %s %s" % 
+                         (orig,new,path_name) )
+        status, output = getstatusoutput(command)
+        if status != 0:
+            tty.warn('failed writing id for %s.' % path_name)
+            tty.warn(output)
+
+    os.chmod(path_name, st.st_mode)
+    return
+
+
+
+def get_filetype(path_name):
+    """
+    Check the output of the file command for given string.
+    """
+    file_command = "LC_ALL=C file -b -h \"%s\"" % path_name
+    status, output = getstatusoutput(file_command)
+    if status != 0:
+        tty.warn('getting filetype of "%s" failed' % path_name)
+        return None
+    return output.strip()
+
+
+def modify_elf_object(path_name, orig_rpath, new_rpath, patchelf_executable):
+    """
+    Replace RPATH's in given elf object
+    """
+    st = os.stat(path_name)
+    wmode = os.access(path_name, os.W_OK)
+    if not wmode:
+        os.chmod(path_name, st.st_mode | stat.S_IWUSR)
+    if platform.system() == 'Linux':
+        new_joined = ':'.join(new_rpath)
+        command = "%s --force-rpath --set-rpath '%s' '%s'" % \
+            (patchelf_executable, new_joined, path_name)
+        status, output = getstatusoutput(command)
+        if status != 0:
+            tty.warn('failed writing rpath for %s.' % path_name)
+    else:
+        tty.die('relocation not supported for this platform')
+    os.chmod(path_name, st.st_mode)
+
+
+def needs_binary_relocation(filetype):
+    """
+    Check whether the given filetype is a binary that may need relocation.
+    """
+    retval = False
+    if "relocatable" in filetype:
+        return False
+    if platform.system() == 'Darwin':
+        return ('Mach-O' in filetype)
+    elif platform.system() == 'Linux':
+        return ('ELF' in filetype)
+    else:
+        tty.die("Relocation not implemented for %s" % platform.system())
+    return retval
+
+
+def needs_text_relocation(filetype):
+    """
+    Check whether the given filetype needs relocation.
+    """
+    return ("text" in filetype)
+
+
+def relocate_binary(path_name, old_dir, new_dir, patchelf_executable):
+    """
+    Change RPATHs in given elf or mach-o file
+    """
+    if platform.system() == 'Darwin':
+        modify_macho_object(path_name, old_dir, new_dir)
+    elif platform.system() == 'Linux':
+        orig_rpaths = get_existing_elf_rpaths(path_name, patchelf_executable)
+        new_rpaths  = substitute_rpath(orig_rpaths, old_dir, new_dir)
+        modify_elf_object(path_name, orig_rpaths, new_rpaths, patchelf_executable)
+    else:
+        tty.die("Relocation not implemented for %s" % platform.system())
+
+
+def relocate_text(path_name, old_dir, new_dir):
+    """
+    Replace old path with new path in text files
+    """
+    os.system("LC_ALL=C sed -i -e \"s#%s#%s#g\" \"%s\""
+              % (old_dir, new_dir, path_name))
```

Comment:
This should probably use `filter_file` (which spack packages use) from `llnl.util.filesystem`.